docs/articles/concepts/joins.md

```diff
@@ -1,3 +1,775 @@
+downstream.
+
+| Key | Name Value | Email Value |
+| --- | ---------- | ----------- |
+| 456 | `"Hive"` | |
+
+```
+('123', ('Bee', 'bee@bytewax.io'))
+```
+
+Let's continue. We then see the second email come through and clear
+that row.
+
+| Key | Name Value | Email Value |
+| --- | ---------- | ----------- |
+
+```
+('456', ('Hive', 'hive@bytewax.io'))
+```
+
+So note, now that the table is empty, when we see the email update for
+the Bee, there's no name state. So nothing is emitted!
+
+| Key | Name Value | Email Value |
+| --- | ---------- | ----------- |
+| 123 | | `"bee@bytewax.io"` |
+
+Hopefully this helps clarify how basic streaming joins work. Realizing
+that the
+[`join`](/apidocs/bytewax.operators/index#bytewax.operators.join)
+operator only keeps the state around while it is waiting for all sides
+to complete is the trick to remember.
+
+### Running Joins
+
+The streaming join semantics are useful in some cases, but on infinite
+data you could imagine other join semantics. A **running join** emits
+downstream the values for all sides of the join whenever any value
+comes in, _and keeps the state around_. This is similar to a _full
+outer join_ in SQL. The way to enable a running join with the
+[`join`](/apidocs/bytewax.operators/index#bytewax.operators.join)
+operator is to set `running=True`.
+
+Let's review what the dataflow would look like then:
+
+```python
+flow = Dataflow("join_eg")
+
+names_l = [
+    {"user_id": 123, "name": "Bee"},
+    {"user_id": 456, "name": "Hive"},
+]
+names = op.input("names", flow, TestingSource(names_l))
+
+emails_l = [
+    {"user_id": 123, "email": "bee@bytewax.io"},
+    {"user_id": 456, "email": "hive@bytewax.io"},
+    {"user_id": 123, "email": "queen@bytewax.io"},
+]
+emails = op.input("emails", flow, TestingSource(emails_l))
+
+keyed_names = op.map("key_names", names, lambda x: (str(x["user_id"]), x["name"]))
+keyed_emails = op.map("key_emails", emails, lambda x: (str(x["user_id"]), x["email"]))
+
+joined = op.join("join", keyed_names, keyed_emails, running=True)
+```
+
+Now let's run the dataflow again an inspect the output.
+
+```python
+op.inspect("check_join", joined)
+
+bytewax.testing.run_main(flow)
+```
+
+Here's what we get. Let's visualize the progress and outputs of the
+[`join`](/apidocs/bytewax.operators/index#bytewax.operators.join)
+state table again but with `running=True`.
+
+```{testoutput}
+join_eg.check_join: ('123', ('Bee', None))
+join_eg.check_join: ('123', ('Bee', 'bee@bytewax.io'))
+join_eg.check_join: ('456', ('Hive', None))
+join_eg.check_join: ('456', ('Hive', 'hive@bytewax.io'))
+join_eg.check_join: ('123', ('Bee', 'queen@bytewax.io'))
+```
+
+First the table starts empty.
+
+| Key | Name Value | Email Value |
+| --- | ---------- | ----------- |
+
+Then we see the first name value. Since the table was updated in any
+way, it emits the current values for that key and keeps them in the
+table! If it doesn't know a value yet, it fills it in as `None`.
```

Comment:
Let's include right here a short suggestion on what the reader can do with such `None` values, even if we think it's obvious that it should be a separate step downstream, we should explicitly encourage the user to do it this way.